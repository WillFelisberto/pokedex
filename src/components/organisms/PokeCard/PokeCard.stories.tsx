import type { Meta, StoryObj } from '@storybook/react';

import { PokeCard } from './PokeCard';

const meta: Meta<typeof PokeCard> = {
  title: 'Organisms/PokeCard', // Categoria e nome do componente
  component: PokeCard, // Componente sendo documentado
  tags: ['autodocs'], // Adiciona documentação automática
  args: {
    // Valores padrão para as props
    pokemon: {
      baseStats: {
        hp: 45,
        attack: 49,
        defense: 49,
        specialAttack: 65,
        specialDefense: 65,
        speed: 45
      },
      japaneseName: 'イビス',
      region: 'Kanto',
      id: 2,
      name: 'ivysaur',
      sprites: {
        front_default:
          'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/2.png'
      },
      types: ['grass', 'poison'],
      category: 'ivysaur',
      weight: 130,
      height: 10,
      weaknesses: ['flying', 'poison', 'bug', 'fire', 'ice'],
      evolutions: [
        {
          id: 1,
          name: 'bulbasaur',
          url: 'https://pokeapi.co/api/v2/pokemon/1/',
          sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png',
          types: ['grass', 'poison'],
          level: null
        },
        {
          id: 2,
          name: 'ivysaur',
          url: 'https://pokeapi.co/api/v2/pokemon/2/',
          sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png',
          types: ['grass', 'poison'],
          level: 16
        },
        {
          id: 3,
          name: 'venusaur',
          url: 'https://pokeapi.co/api/v2/pokemon/3/',
          sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/3.png',
          types: ['grass', 'poison'],
          level: 32
        }
      ],
      abilities: [
        { name: 'overgrow', is_hidden: false },
        { name: 'chlorophyll', is_hidden: true }
      ],
      genderRatio: {
        male: 87.5,
        female: 12.5
      }
    }
  }
};

export default meta;

// Tipo para as histórias
type Story = StoryObj<typeof PokeCard>;

// História padrão
export const Default: Story = {};

// História sem evoluções
export const SemEvolucoes: Story = {
  args: {
    pokemon: {
      baseStats: {
        hp: 45,
        attack: 49,
        defense: 49,
        specialAttack: 65,
        specialDefense: 65,
        speed: 45
      },
      japaneseName: 'イビス',
      region: 'Kanto',
      ...(meta.args?.pokemon ?? {}),
      evolutions: [], // Remove as evoluções
      id: 2, // Ensure id is defined
      name: 'ivysaur', // Ensure name is defined
      sprites: {
        front_default:
          'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/2.png'
      },
      types: ['grass', 'poison'],
      category: 'ivysaur',
      weight: 130,
      height: 10,
      weaknesses: ['flying', 'poison', 'bug', 'fire', 'ice'],
      abilities: [
        { name: 'overgrow', is_hidden: false },
        { name: 'chlorophyll', is_hidden: true }
      ],
      genderRatio: {
        male: 87.5,
        female: 12.5
      }
    }
  }
};

// História com um Pokémon diferente
export const Charmander: Story = {
  args: {
    pokemon: {
      baseStats: {
        hp: 45,
        attack: 49,
        defense: 49,
        specialAttack: 65,
        specialDefense: 65,
        speed: 45
      },
      japaneseName: 'イビス',
      region: 'Kanto',
      id: 4,
      name: 'charmander',
      sprites: {
        front_default:
          'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/4.png'
      },
      types: ['fire'],
      category: 'charmander',
      weight: 85,
      height: 6,
      weaknesses: ['water', 'ground', 'rock'],
      evolutions: [
        {
          id: 4,
          name: 'charmander',
          url: 'https://pokeapi.co/api/v2/pokemon/4/',
          sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png',
          types: ['fire'],
          level: null
        },
        {
          id: 5,
          name: 'charmeleon',
          url: 'https://pokeapi.co/api/v2/pokemon/5/',
          sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/5.png',
          types: ['fire'],
          level: 16
        },
        {
          id: 6,
          name: 'charizard',
          url: 'https://pokeapi.co/api/v2/pokemon/6/',
          sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/6.png',
          types: ['fire', 'flying'],
          level: 36
        }
      ],
      abilities: [
        { name: 'blaze', is_hidden: false },
        { name: 'solar-power', is_hidden: true }
      ],
      genderRatio: {
        male: 87.5,
        female: 12.5
      }
    }
  }
};
