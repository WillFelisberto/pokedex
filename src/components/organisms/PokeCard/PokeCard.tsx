import Link from 'next/link';
import React from 'react';

import { PokeBadge } from '@/components/atoms/PokeBadge';
import { PokemonHeader } from '@/components/atoms/PokemonHeader';
import { PokemonStatus } from '@/components/molecules/PokemonStatus';

import { PokemonType } from '@/lib/utils/poketypes';
import { formatPokemonId } from '@/lib/utils/StatusConvert';

export interface PokemonProps {
  id: number;
  name: string;
  japaneseName?: string;
  region: string;
  sprites: {
    front_default: string;
  };
  types: PokemonType[];
  category: string;
  weight: number;
  height: number;
  weaknesses: PokemonType[];
  evolutions: {
    id: number;
    name: string;
    url: string;
    sprite: string;
    types: PokemonType[];
    level: number | null;
  }[];
  abilities: { name: string; is_hidden: boolean }[];
  genderRatio: {
    male: number;
    female: number;
  };
  baseStats: {
    hp: number;
    attack: number;
    defense: number;
    specialAttack: number;
    specialDefense: number;
    speed: number;
  };
}

export const PokeCard: React.FC<{ pokemon: PokemonProps }> = ({ pokemon }) => {
  return (
    <div
      data-testid="poke-card"
      key={pokemon.id}
      className="bg-white min-w-[360px] max-w-[360px] p-2 rounded-lg text-center flex flex-col gap-4"
    >
      <Link href={`/pokemon/${pokemon.id}`}>
        <PokemonHeader
          name={pokemon.name}
          spriteUrl={pokemon.sprites.front_default}
          types={pokemon.types}
        />
        <div className="p-2 text-left">
          <h2 className="text-black font-bold capitalize text-3xl" data-testid="pokemon-name">
            {pokemon.name}
          </h2>
          <p
            className="text-base text-gray-700"
            data-testid="pokemon-id"
          >{`NÂº${formatPokemonId(pokemon.id)}`}</p>
        </div>
        <div className="p-2 flex gap-2" data-testid="pokemon-types">
          {pokemon.types.map((type) => (
            <PokeBadge key={`${type}-${pokemon.id}`} type={type} />
          ))}
        </div>
        <div className="p-2 w-full flex gap-[20px]">
          <PokemonStatus height={pokemon.height} weight={pokemon.weight} />
        </div>
      </Link>
    </div>
  );
};
