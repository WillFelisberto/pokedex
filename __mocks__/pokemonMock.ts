import { PokemonProps } from '@/components/organisms/PokeCard';

export const mockPokemon: PokemonProps = {
  id: 2,
  name: 'ivysaur',
  sprites: {
    front_default:
      'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/2.png'
  },
  types: ['grass', 'poison'],
  category: 'ivysaur',
  weight: 130,
  height: 10,
  weaknesses: ['flying', 'poison', 'bug', 'fire', 'ice'],
  evolutions: [
    {
      id: 1,
      name: 'bulbasaur',
      url: 'https://pokeapi.co/api/v2/pokemon/1/',
      sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png',
      types: ['grass', 'poison'],
      level: null
    },
    {
      id: 2,
      name: 'ivysaur',
      url: 'https://pokeapi.co/api/v2/pokemon/2/',
      sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png',
      types: ['grass', 'poison'],
      level: 16
    },
    {
      id: 3,
      name: 'venusaur',
      url: 'https://pokeapi.co/api/v2/pokemon/3/',
      sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/3.png',
      types: ['grass', 'poison'],
      level: 32
    }
  ],
  abilities: [
    { name: 'overgrow', is_hidden: false },
    { name: 'chlorophyll', is_hidden: true }
  ],
  genderRatio: {
    male: 87.5,
    female: 12.5
  }
};
