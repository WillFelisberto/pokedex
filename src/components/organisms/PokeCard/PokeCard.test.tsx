import { mockPokemon } from '@/mocks/pokemonMock';
import { render, screen } from '@/tests/test-utils';
import React from 'react';

import { PokeCard } from './PokeCard';

jest.mock('next/navigation', () => ({
  useSearchParams: () => ({
    get: jest.fn().mockReturnValue('1') // Simula "page=1" na URL
  })
}));

describe('PokeCard', () => {
  it('Should render the component', () => {
    const { container } = render(<PokeCard pokemon={mockPokemon} />);
    expect(screen.getByTestId('poke-card')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('Should render the pokemon name', () => {
    render(<PokeCard pokemon={mockPokemon} />);
    const pokemonName = screen.getByTestId('pokemon-name');
    expect(pokemonName).toBeInTheDocument();
    expect(pokemonName).toHaveTextContent(/ivysaur/i);
  });

  it('Should render the pokemon number', () => {
    render(<PokeCard pokemon={mockPokemon} />);

    const pokemonId = screen.getByTestId('pokemon-id');
    expect(pokemonId).toBeInTheDocument();
    expect(pokemonId).toHaveTextContent(/NÂº002/i);
  });

  it('Should render the pokemon types', () => {
    render(<PokeCard pokemon={mockPokemon} />);

    const pokemonTypes = screen.getByTestId('pokemon-types');
    expect(pokemonTypes).toBeInTheDocument();
  });
});
