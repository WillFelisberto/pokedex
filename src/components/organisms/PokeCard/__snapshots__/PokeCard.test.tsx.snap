// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokeCard Should render the component 1`] = `
<div>
  <div
    class="bg-white min-w-[340px] max-w-[340px] p-2 rounded-lg text-center flex flex-col gap-4"
    data-testid="poke-card"
  >
    <a
      href="/pokemon/2"
    >
      <div
        class="relative w-full h-[280px] flex flex-col items-center justify-center rounded-b-3xl overflow-hidden"
      >
        <div
          class="relative w-full h-full flex items-center justify-center"
          data-testid="pokemon-header"
          style="clip-path: ellipse(60% 65% at 50% 30%);"
        >
          <div
            class="absolute inset-0 flex items-center justify-center"
          >
            <div
              class="absolute inset-0 w-1/2 bg-gradient-to-l from-white/15 to-transparent"
            />
            <div
              class="relative"
            >
              <img
                alt="grass"
                height="140"
                src="/icons/png/filled/grass.png"
                style="mask-image: linear-gradient(135deg, rgba(0 0 0 / 50%), rgba(0,0,0,0)); filter: brightness(10);"
                width="140"
              />
            </div>
          </div>
        </div>
        <div
          class="  w-full flex justify-center mt-[5rem] absolute"
        >
          <img
            alt="ivysaur"
            class="z-10"
            height="215"
            src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/2.png"
            width="215"
          />
        </div>
      </div>
      <div
        class="p-2 text-left"
      >
        <h2
          class="text-black font-bold capitalize text-3xl"
          data-testid="pokemon-name"
        >
          ivysaur
        </h2>
        <p
          class="text-base text-gray-700"
          data-testid="pokemon-id"
        >
          NÂº002
        </p>
      </div>
      <div
        class="p-2 flex gap-2"
        data-testid="pokemon-types"
      >
        <div
          aria-label="grass"
          class="flex w-fit items-center gap-2 text-white px-3 py-2 rounded-full bg-pokemon-grass"
          data-testid="poke-badge"
        >
          <div
            class="w-6 h-6 flex items-center justify-center rounded-full overflow-visible bg-white"
          >
            <img
              alt="grass"
              height="13"
              src="/icons/png/filled/grass.png"
              width="13"
            />
          </div>
          <span
            class="font-semibold text-sm"
          >
            Grass
          </span>
        </div>
        <div
          aria-label="poison"
          class="flex w-fit items-center gap-2 text-white px-3 py-2 rounded-full bg-pokemon-poison"
          data-testid="poke-badge"
        >
          <div
            class="w-6 h-6 flex items-center justify-center rounded-full overflow-visible bg-white"
          >
            <img
              alt="poison"
              height="13"
              src="/icons/png/filled/poison.png"
              width="13"
            />
          </div>
          <span
            class="font-semibold text-sm"
          >
            Poison
          </span>
        </div>
      </div>
      <div
        class="p-2 w-full flex gap-[20px]"
      >
        <div
          class="grid grid-cols-2 w-full gap-[20px]"
        >
          <div
            class="flex flex-col gap-[4.5px] items-left   rounded-lg  "
          >
            <div
              class="flex items-end gap-[8.5px] h-[16px] text-gray-500 text-xs font-semibold uppercase"
            >
              <img
                alt="weight"
                height="18"
                src="/icons/svg/outlined/weight.svg"
                width="18"
              />
              <span>
                Weight
              </span>
            </div>
            <div
              class="mt-1 font-(--font-poppins) h-[42px]  w-full py-[8px] bg-white rounded-[15px] text-gray-800 font-bold text-lg text-center border border-[#E5E5E5]"
            >
              13 kg
            </div>
          </div>
          <div
            class="flex flex-col gap-[4.5px] items-left   rounded-lg  "
          >
            <div
              class="flex items-end gap-[8.5px] h-[16px] text-gray-500 text-xs font-semibold uppercase"
            >
              <img
                alt="height"
                height="15"
                src="/icons/svg/outlined/height.svg"
                width="15"
              />
              <span>
                Height
              </span>
            </div>
            <div
              class="mt-1 font-(--font-poppins) h-[42px]  w-full py-[8px] bg-white rounded-[15px] text-gray-800 font-bold text-lg text-center border border-[#E5E5E5]"
            >
              1 m
            </div>
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;
